Access single file or folder from IStorageFolder by name

nickodei opened the pull request:
## What does the pull request do?

This PR addresses #17276. It allows to access a single file or folder **by name** given you have already a folder (`IStorageFolder`).


## What is the current behavior?
You can currently archive this by iteration over the whole folder with `GetItemsAsync()` and pick the file with the given name. This has performance considerations and therefore, when applicable, the implementations tries to access the file directly.


## What is the updated/expected behavior with this PR?
When you get a folder (`IStorageFolder`) e.g through a file-picker, you can access the child file/folder by name directly, returning `IStorageFile` or `IStorageFolder` depending on the called method.


## How was the solution implemented (if it's not obvious)?
The interface `IStorageFolder` was extended as follows:
```
interface IStorageFolder
{
     Task<IStorageFolder> GetFolderAsync(string name);
     Task<IStorageFile> GetFileAsync(string name);
}
```

Therefore, an implementation for the following platforms was provided:

- Android (DocumentsContract and the query method)
- Browser (FileSystemDirectoryHandle)
- Desktop (.NET Path.Combine)
- IOS (.NET Path.Combine => NSUrl)


## Testing

Tested it on following platforms:

- [x] Desktop 
- [x] Android 
- [x] Browser 
- [ ] IOS (dont have an IOS device)

Extended the ControlCatalog for testing:

![image](https://github.com/user-attachments/assets/e666e668-10fb-4c43-9083-5238373acb6c)
 

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
No breaking changes, this only extends the public interface

## Fixed issues
Fixes #17276

nickodei commented:
I was not able to have an Android-Implementation that does not iterate trough the whole folder to find the first Item with the given name. The other thing is, that I don't have an IOS device to test the IOS-implementation.

nickodei commented:
Questions:

- Should I return `IStorageFolder` or `IStorageFolder?` ? => When nullable, which Excetions should be catched and which should be thrown (e.g getDirectoryHandle throws 4 different Exceptions. Should I catch all of them and return null?)
- Can I write tests for it or is it enough that I manually tested it?

MrJul commented:
> * Should I return `IStorageFolder` or `IStorageFolder?` ? => When nullable, which Excetions should be catched and which should be thrown (e.g getDirectoryHandle throws 4 different Exceptions. Should I catch all of them and return null?)

We'll discuss that in an API review meeting soon and let you know, but I think the return type should be nullable. After all, even the `Create{File/Folder}Async` methods can return `null`.

> * Can I write tests for it or is it enough that I manually tested it?

Platform specific code is unfortunately hard to test, and we don't currently have integration tests for storage, so manual testing should be ok here.

> * How to I resolve the build error?

Adding new members to an interface is a breaking change, which is what the CI build reports here. However, `IStorageFolder` is marked with `NotClientImplementable` so we exceptionally allow breaking changes here as only platforms should implement it.

You should run `nuke ValidateApiDiff` to generate an API suppression file (but you might want to wait until we've properly reviewed the new API).

MrJul commented:
Diff for API review:

```diff
 namespace Avalonia.Platform.Storage {
 {
     [NotClientImplementable]
     public interface IStorageFolder : IStorageItem
     {
+        System.Threading.Tasks.Task<IStorageFile?> GetFileAsync(System.String name);
+        System.Threading.Tasks.Task<IStorageFolder?> GetFolderAsync(System.String name);
     }
 }
```

nickodei commented:
Thank you for your response. I'm still down to finish this MR so I will wait until the API review.

MrJul commented:
Notes from the API review meeting:

The API shape is good.

However, all platforms should behave the same, as much as possible.
For example, on the browser, there's an exception explicitly thrown at the start of the method (wrong copy/paste?), which should return null instead.

nickodei committed:
added methods to IStorageFolder to access a single file or folder by name

nickodei commented:
Thanks. I incorporated the browser change and updated the api-diff. It seems a bit weird because throwing vs. returning null is not everywhere strictly enforced when using the nullable operator. Should I just catch all Exceptions and return null? Event in the browser example, while I return null (if it is not a directory), there is still the potential to throw an Exception (wrong access, etc. ). But for the most part is shouldn't.

At @MrJul: Are you able to review my changes?

nickodei committed:
relaxed file/folder by name browser acces to throw less errors

nickodei commented:
> Other implementations should match this behavior. Return `null` if the file can't be listed.
> `Bcl`, `Android` and `iOS` all seem correct.
> `Browser` needs swallowing the unauthorized access exception in the current catch block.

Got it and also re-tested the browser implementation. 
There where also some other JS-errors throws that I needed to suppress (NotFound and TypeMissmatch) to match the behavior of the other implementation.

MrJul reviewed (approved):
LGTM, thank you!
